src/extensions/compression/deflate.rs

```diff
@@ -0,0 +1,289 @@
+use std::io::Write;
+
+use flate2::{Compress, Compression, Decompress, FlushCompress, FlushDecompress, Status};
+use http::HeaderValue;
+use thiserror::Error;
+
+use crate::{
+    extensions::{self, Param},
+    protocol::Role,
+};
+
+const PER_MESSAGE_DEFLATE: &str = "permessage-deflate";
+const CLIENT_NO_CONTEXT_TAKEOVER: &str = "client_no_context_takeover";
+const SERVER_NO_CONTEXT_TAKEOVER: &str = "server_no_context_takeover";
+const CLIENT_MAX_WINDOW_BITS: &str = "client_max_window_bits";
+const SERVER_MAX_WINDOW_BITS: &str = "server_max_window_bits";
+
+const TRAILER: [u8; 4] = [0x00, 0x00, 0xff, 0xff];
+
+/// Error from `permessage-deflate` extension.
+#[derive(Debug, Error)]
+pub enum DeflateError {
+    /// Compress failed
+    #[error("failed to compress: {0}")]
+    Compress(std::io::Error),
+    /// Decompress failed
+    #[error("failed to decompress: {0}")]
+    Decompress(std::io::Error),
+}
+
+// Parameters `server_max_window_bits` and `client_max_window_bits` are not supported for now
+// because custom window size requires `flate2/zlib` feature.
+// TODO Configs for how the server accepts these offers.
+/// Configurations for `permessage-deflate` Per-Message Compression Extension.
+#[derive(Clone, Copy, Debug)]
+pub struct DeflateConfig {
+    /// Compression level.
+    pub compression: Compression,
+    /// Request the peer server not to use context takeover.
+    pub server_no_context_takeover: bool,
+    /// Hint that context takeover is not used.
+    pub client_no_context_takeover: bool,
+}
+
+impl Default for DeflateConfig {
+    fn default() -> Self {
+        Self {
+            compression: Compression::default(),
+            server_no_context_takeover: false,
+            client_no_context_takeover: false,
+        }
+    }
+}
+
+impl DeflateConfig {
+    pub(crate) fn name(&self) -> &str {
+        PER_MESSAGE_DEFLATE
+    }
+
+    /// Value for `Sec-WebSocket-Extensions` request header.
+    pub(crate) fn negotiation_offers(&self) -> HeaderValue {
```

Comment:
I actually had this named `generate_offers` at first. It's plural because I was thinking maybe we should support stacked offers.

The name "negotiation offers" was taken from the RFC, but I don't really like it. I'll change it to `generate_offer`.